Add backup and restore guide (#194)
scalar-labs/scalar-kubernetes: docs/BackupNoSQL.md

# Back up a NoSQL database in a Kubernetes environment

> This guide explains how to create a transactionally consistent backup of managed databases that ScalarDB or ScalarDL uses in a Kubernetes environment. Please note that, when using a NoSQL database or multiple databases, you **must** pause ScalarDB or ScalarDL when creating a transactionally consistent backup.

Reviewer: ```suggestion
This guide explains how to create a transactionally consistent backup of managed databases that ScalarDB or ScalarDL uses in a Kubernetes environment. Please note that, when using a NoSQL database or multiple databases, you **must** pause ScalarDB or ScalarDL to create a transactionally consistent backup.
```
Isn't it unnatural to use `when` multiple times in one sentence? (> @josh-wong )

Reviewer 2: My mistake. Your change (`to create`) is better in this case.

Author: Thank you for your confirmation! I will apply the suggestion.

For details on how ScalarDB backs up databases, see [A Guide on How to Backup and Restore Databases Used Through ScalarDB](https://github.com/scalar-labs/scalardb/blob/master/docs/backup-restore.md).

> In this guide, we assume that you are using the automatic backup and point-in-time recovery (PITR) features. Therefore, we must create a period where there are no ongoing transactions for restoration. You can then restore data to that specific period by using PITR. If you restore data to a time without creating a period where there are no ongoing transactions, the restored data could be transactionally inconsistent, causing ScalarDB or Scalar DL to not work properly with the data.

Author: Ah, I see. Indeed, the automatic backup and PIRT are almost one feature.

I will apply your suggestion. Thank you!

7. Check the pod status after performing the backup. You must check the following four points by using the `kubectl get pod` command after the backup operation is completed.
   * **The number of ScalarDB or ScalarDL pods.** Confirm this number matches the number of pods that you wrote down before performing the backup.
   * **The ScalarDB or ScalarDL pod names in the `NAME` column.** Confirm the names match the pod names that you wrote down before performing the backup.
   * **The ScalarDB or ScalarDL pod status is `Running` in the `STATUS` column.**
   * **The restart count of each pod in the `RESTARTS` column.** Confirm the counts match the restart counts that you wrote down before performing the backup
   
   **If any of the two values are different, you must retry the backup operation from the beginning.** The reason for the different values may be caused by some pods being added or restarted while performing the backup. In such case, those pods will run in the `unpause` state. Pods in the `unpause` state will cause the backup data to be transactionally inconsistent.

> 8. **(Amazon DynamoDB only)** If you use the PITR feature in DynamoDB, you will need to perform additional steps to create a backup because the feature restores data with another name table by using PITR. For details on the additional steps after creating the exact period in which you can restore the data, please see [Restore databases in a Kubernetes environment](./RestoreDatabase.md#amazon-dynamodb).

Reviewer: ```suggestion
8. **(Amazon DynamoDB only)** If you use the PITR feature of DynamoDB, you will need to perform additional steps to create a backup because the feature restores data with another name table by using PITR. For details on the additional steps after creating the exact period in which you can restore the data, please see [Restore databases in a Kubernetes environment](./RestoreDatabase.md#amazon-dynamodb).
```

> # Backup NoSQL database guide

Reviewer: ```suggestion
# A Guide on Taking Backups of NoSQL Databases on a Kubernetes Environment
```

We might need some naming rules (e.g., capitalization, strucfor this kind of guide.
@josh-wong Thought?

Reviewer 2: Regarding title and heading capitalization, I'd like us to follow similar capitalization guidelines mentioned in the [Microsoft Style Guide ("Formatting headings")](https://learn.microsoft.com/en-us/style-guide/scannable-content/headings#formatting-headings) and [Google Developer style guide ("Headings and title format")](https://developers.google.com/style/headings#heading-and-title-format). Both style guides recommend using sentence-style capitalization.

Regarding title structure, I'd like us to follow a similar style mentioned in the [Google Developer style guide ("Heading and title text")](https://developers.google.com/style/headings#heading-and-title-text). This style guide shows how to write tasked-based and non-task-based headings.

I'll be sure to add both of the above to our style guide.

Reviewer 2: Following the guidance in [Google Developer Documentation Style Guide ("Heading and title text")](https://developers.google.com/style/headings#heading-and-title-text), how about we use the following?:

```suggestion
# Back up NoSQL databases in a Kubernetes environment
```

> If you use a NoSQL database or multiple databases, you must **pause** the Scalar products to create a transactionally-consistent backup. This guide explains concrete steps of how to create the transactionally-consistent backup of managed databases with the Scalar products on a Kubernetes environment. Please refer to the following document if you want to know more details on the backup way of Scalar products.

Reviewer: Is there a particular reason for saying `the Scalar products` here and saying `ScalarDB/DL` in docs/BackupRestoreGuide.md?
I think we should use consistent naming as much as possible if we talk about the same thing.

Reviewer 2: To reduce ambiguity, I've specified ScalarDB and ScalarDL as the target products.

```suggestion
If you are using a NoSQL database or multiple databases, you must pause the Scalar products to create a transactionally consistent backup. This guide explains how to create a transactionally consistent backup of managed databases by using ScalarDB or ScalarDL in a Kubernetes environment. 

For details on how ScalarDB backs up databases, see [A Guide on How to Backup and Restore Databases Used Through ScalarDB](https://github.com/scalar-labs/scalardb/blob/master/docs/backup-restore.md).
```

Author: I feel `databases that use ScalarDB or ScalarDL` is `The database utilizes some ScalarDB/ScalarDL feature`.
In other words, I feel that sentence means the ScalarDB/ScalarDL are extensions for some databases like the following image.
```
[Client] -------> [Database (with ScalarDB)]

            +---> [Database (with ScalarDL Ledger)]
            |
[Client] ---+
            |
            +---> [Database (with ScalarDL Auditor)]
```

However, in fact, the system architecture is the following.
```
[Client] -------> [ScalarDB] -----------> [Backend DB]

            +---> [ScalarDL Ledger] ----> [Backend DB]
            |
[Client] ---+
            |
            +---> [ScalarDL Auditor] ---> [Backend DB]
```

The ScalarDB and ScalarDL utilize the database as a `Backend DB` to store some data of ScalarDB or ScalarDL.
So, I think it would be better to describe `databases that are used by ScalarDB or ScalarDL` instead of `databases that use ScalarDB or ScalarDL`.

What do you think?

> 
> * [A Guide on How to Backup and Restore Databases Used Through ScalarDB](https://github.com/scalar-labs/scalardb/blob/master/docs/backup-restore.md)

Reviewer 2: We can delete these lines since I added the link to the previous sentence.

```suggestion
```

> In this guide, we assume that using the automatic backup and PITR feature. So, we create a **period** where there are no ongoing transactions for restoration. You can restore data using PITR to the point-in-time in this **period**. Otherwise (if you restore data to point-in-time out of this **period**), the restored data becomes inconsistent data and Scalar products don't work properly.

Reviewer: ```suggestion
In this guide, we assume that using automatic backup and PITR features. So, we create a **period** where there are no ongoing transactions. You can restore data using PITR to some point in time in this **period**. If you restore data to some point in time out of this **period**), the restored data could be transactionally-inconsistent; Scalar products won't work properly with the data.
```

Reviewer 2: ```suggestion
In this guide, we assume that you are using the automatic backup and point-in-time recovery (PITR) features. Therefore, we must create a period where there are no ongoing transactions for restoration. You can then restore data to that specific period by using PITR. If you restore data to a time without creating a period where there are no ongoing transactions, the restored data could be transactionally inconsistent, causing ScalarDB or Scalar DL to not work properly with the data.
```

> ## Backup operations (create the **period** for restoration)

Author: We will create a new tool for creating the restoration period to cover the rare case issues.
However, it takes a bit of time to create, so I created this backup document based on the old (current) way at this time.
I will update this document after the new tool will be released.

> ## Backup operations (create the **period** for restoration)

Reviewer 2: ```suggestion
## Back up the data by creating a period for data restoration
```

1. Unpause Scalar Product pods using `scalar-admin`.

   Please refer to the [How to use scalar-admin on a Kubernetes environment](./BackupNoSQL.md#how-to-use-scalar-admin-on-a-kubernetes-environment) section in this document for more details on the unpause operation.

1. Check the unpause started time.

   You must note the unpause started time to decide the concrete **period** that you can restore data using the PITR feature.

1. Check the pod status after backup.

   You must check the following four points using the `kubectl get pod` command after you finish the backup operation.
   * The number of Scalar product pods (You must compare this number with the number before the backup).
   * The Scalar product pod names in the `NAME` column (You must compare these pod names with the pod names before the backup).
   * The Scalar product pod status is `Running` in the `STATUS` column.
   * The restart counts of each pod in the `RESTARTS` column (You must compare these restart counts with the restart counts before the backup).

>    **If the two values are different, you must re-try the backup operation all over again.** This is because, if some pods are added or restarted, those pods run with a `unpause` state. The `unpause` state pods cause data inconsistency in the backup data.

Author: @brfrn169  
Thank you for your review! (and sorry for the late response...)

I think the steps you pointed out are an ideal way, and I will create a new tool based on those steps.
However, in the current document, users deploy the `scalar-admin` pod, and it returns the **unpause started time** in the stdout.
It is a bit confusing and I think there is a possibility that users will make misunderstand the unpause started time as follows.

1. Check the pod status before backup.
1. Pause Scalar Product pods using `scalar-admin`.
1. Check the pause completed time. <- Users can do this step based on the output of `scalar-admin` pod.
1. Wait for about 10 seconds.
1. Check the unpause started time. <- Users cannot use the output of `scalar-admin` pod.  They need to check this time using their own clock.
1. Check the pod status after backup.
1. === Here is a time lag ===
1. Unpause Scalar Product pods using `scalar-admin`. <- The `scalar-admin` pod outputs the **unpause started time**, but users cannot use this time to decide the **period** since this time is later than status checking.

In these steps, users need to decide the **period** based on the `step 3` and the `step 5`.
In other words, there is a possibility that the pods are created or restarted at the point of `step 7` in the above steps. This causes backup data inconsistency.
However, when users run the unpause requests at `step 8`, the `scalar-admin` pod outputs **unpause started time** in the stdout.
If users pick up this time as a unpause started time (this is a wrong decision), there is a possibility that the backup data inconsistency occurs.
This output is confusing, but there is no way to disable outputting **unpause started time** in the stdout.

As the above, to avoid making misunderstandings of users, I think it is better to run unpause requests before checking pod status and check the unpause started time based on the output of `scalar-admin` pod.
I think this step also can keep the consistency of backup data since the **pause completed time** and **unpause started time** (i.e., the **period**) exist between the two status checking operations.

What do you think?

>    **If the two values are different, you must re-try the backup operation all over again.** This is because, if some pods are added or restarted, those pods run with a `unpause` state. The `unpause` state pods cause data inconsistency in the backup data.

Reviewer 3: Sounds good. Thank you for the explanation!

> 1. Check the pod status before backup.
> 
>    You must check the following four points using the `kubectl get pod` command before you start the backup operation.
>    * The number of Scalar product pods (You must compare this number with the number after the backup).
>    * The Scalar product pod names in the `NAME` column (You must compare these pod names with the pod names after the backup).
>    * The Scalar product pod status is `Running` in the `STATUS` column.
>    * The restart counts of each pod in the `RESTARTS` column (You must compare these restart counts with the restart counts after the backup).
> 
> 1. Pause Scalar Product pods using `scalar-admin`.
> 
>    Please refer to the [How to use scalar-admin on a Kubernetes environment](./BackupNoSQL.md#how-to-use-scalar-admin-on-a-kubernetes-environment) section in this document for more details on the pause operation.
> 
> 1. Check the pause completed time.
> 
>    You must note the pause completed time to decide the concrete **period** that you can restore data using the PITR feature.
> 
> 1. Take a backup using the backup feature of each database.
> 
>    If you enable the automatic backup and PITR feature, all you have to do is wait for about 10 seconds since the managed databases take backup automatically. This 10 seconds period is the **period** that we can restore data by PITR.
> 
> 1. Unpause Scalar Product pods using `scalar-admin`.
> 
>    Please refer to the [How to use scalar-admin on a Kubernetes environment](./BackupNoSQL.md#how-to-use-scalar-admin-on-a-kubernetes-environment) section in this document for more details on the unpause operation.
> 
> 1. Check the unpause started time.
> 
>    You must note the unpause started time to decide the concrete **period** that you can restore data using the PITR feature.
> 
> 1. Check the pod status after backup.
> 
>    You must check the following four points using the `kubectl get pod` command after you finish the backup operation.
>    * The number of Scalar product pods (You must compare this number with the number before the backup).
>    * The Scalar product pod names in the `NAME` column (You must compare these pod names with the pod names before the backup).
>    * The Scalar product pod status is `Running` in the `STATUS` column.
>    * The restart counts of each pod in the `RESTARTS` column (You must compare these restart counts with the restart counts before the backup).
> 
>    **If the two values are different, you must re-try the backup operation all over again.** This is because, if some pods are added or restarted, those pods run with a `unpause` state. The `unpause` state pods cause data inconsistency in the backup data.

Reviewer 2: ```suggestion
1. Check the following four points by running the `kubectl get pod` command before starting the backup operation:
   * **The number of ScalarDB or ScalarDL pods.** Write down the number of pods so that you can compare that number with the number of pods after performing the backup.
   * **The ScalarDB or ScalarDL pod names in the `NAME` column.** Write down the pod names so that you can compare those names with the pod names after performing the backup.
   * **The ScalarDB or ScalarDL pod status is `Running` in the `STATUS` column.** Confirm that the pods are running before proceeding with the backup. You will need to pause the pods in the next step.
   * **The restart count of each pod in the `RESTARTS` column.** Write down the restart count of each pod so that you can compare the count with the restart counts after performing the backup.
2. Pause the ScalarDB or ScalarDL pods by using `scalar-admin`. For details on how to pause the pods, see the [Details on using `scalar-admin`](./BackupNoSQL.md#details-on-using-scalar-admin) section in this guide.
3. Write down the time when the pause started. You will need to refer to that time when restoring the data by using the PITR feature.
4. Back up each database by using the backup feature. If you have enabled the automatic backup and PITR features, the managed databases will be backed up automatically in approximately 10 seconds. This 10-second period is the exact period in which you can restore data by using the PITR feature.
5. Unpause ScalarDB or ScalarDL pods by using `scalar-admin`. For details on how to unpause the pods, see the [Details on using `scalar-admin`](./BackupNoSQL.md#details-on-using-scalar-admin) section in this guide.
6. Check the unpause start time. You must check the unpause start time to confirm the exact period in which you can restore data by using the PITR feature.
7. Check the pod status after performing the backup. You must check the following four points by using the `kubectl get pod` command after the backup operation is completed.
   * **The number of ScalarDB or ScalarDL pods.** Confirm this number matches the number of pods that you wrote down before performing the backup.
   * **The ScalarDB or ScalarDL pod names in the `NAME` column.** Confirm the names match the pod names that you wrote down before performing the backup.
   * **The ScalarDB or ScalarDL pod status is `Running` in the `STATUS` column.** Confirm that the pods are running and not paused.
   * **The restart count of each pod in the `RESTARTS` column.** Confirm the counts match the restart counts that you wrote down before performing the backup
   
   **If any of the two values are different, you must retry the backup operation from the beginning.** The reason for the different values may be caused by some pods being added or restarted while performing the backup. In such case, those pods will run in the `unpause` state. Pods in the `unpause` state will cause the backup data to be transactionally inconsistent.
```

Author: Sorry, I don't know the correct English context, but I would like to confirm the means of `in approximately 10 seconds`.

```
the managed databases will perform back up automatically in approximately 10 seconds
```
Does this sentence mean "database will perform back up automatically **after** approximately 10 seconds" or "the automatic backup **takes** approximately 10 seconds"?
If yes, I think it is a bit different from the actual behavior.
Managed database's automatic backup feature gets database backup in the background process and we can specify the PITR point per second.
So, basically, managed databases take a backup every second.
In other words, it doesn't take 10 seconds for backup.

---

One of the purposes of this 10 seconds period is to avoid mistakes caused by clock skew between the client and the database.
Regarding ScalarDB/ScalarDL backup, we must get a backup from the period that there is no ongoing Tx because ongoing Tx causes data inconsistency.
```
                          Pause completed   Unpause started
                         /                 /
|-----------------------|-----------------|-----------------------|
 <- Ongoing Tx exists -> <- No Tx exist -> <- Ongoing Tx exists ->

 (Ongoing Tx causes data inconsistency)
```

And, we specify the PITR time based on the **Database clock**.
However, the `scalar-admin` command returns pause duration (time) based on the **Client clock**.

If the client clock and database clock are perfectly synchronized, then we don't need the long time duration.
For example, it is enough for about 2 seconds duration as a backup/restoration period as follows.
```
Client (i.e., scalar-admin) clock
+----+----+----+----+----+----+----+----+----+----+
0    1    2    3    4    5    6    7    8    9    10

Database clock
+----+----+----+----+----+----+----+----+----+----+
0    1    2    3    4    5    6    7    8    9    10

          |         |
          |         * Unpause starting at 4
          |
          * Pause started at 2
```
In this case, users can specify the time `3` as a restoration point.

However, there is no `perfectly synchronized clock` in the real world, so we need to take care of the clock skew.
For example, if the clock is off by two seconds between the client and the database as follows, there is a possibility that it causes mistakes (data inconsistency).
```
Client (i.e., scalar-admin) clock
          +----+----+----+----+----+----+----+----+----+----+
          0    1    2    3    4    5    6    7    8    9    10

Database clock
+----+----+----+----+----+----+----+----+----+----+
0    1    2    3    4    5    6    7    8    9    10

                    |         |
                    |         * Unpause starting at 4 from the client perspective
                    |
                    * Pause started at 2 from the client perspective
```
In this case, users see the pause duration that is from 2 to 4 from the output of scalar-admin based on the client clock.
And, maybe, users specify the time `3` as a restoration point in the database console.

However, in fact, the correct pause duration is from 4 to 6 in the database clock.
So, we need to specify the time `5` to specify the `No Tx exist` period.

If users specify the time `3` in the database console based on the client clock, it causes data inconsistency.
The time `3` in the database side is the duration that is the `Ongoing Tx exists`.

---

To avoid the above issue caused by the clock skew, we need to create a sufficiently long period of time (i.e., 10 seconds).

```
Client (i.e., scalar-admin) clock
          +----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+
          0    1    2    3    4    5    6    7    8    9    10   11   12   13   14   15

Database clock
+----+----+----+----+----+----+----+----+----+----+----+----+----+----+----+
0    1    2    3    4    5    6    7    8    9    10   11   12   13   14   15

               |                                                 |
               |                                                 * Unpause starting at 11 from the client perspective
               |
               * Pause started at 1 from the client perspective
```

For example, if users create a pause period as 10 seconds as the above, the pause duration is from 1 to 11 in the client clock.
And, the duration is from 3 to 13 in the database clock.
So, even if they specify the middle time `6` based on the client clock in the database console when they do the PITR, the time `6` is in the `No Tx exist` duration (3 ~ 13) in the database clock.
This doesn't cause data inconsistency.

So, the important thing is not `waiting for the automatic backup feature to complete taking a backup`.
The important thing is `make a sufficiently long period`.
This is why we wait 10 seconds.

---

Sorry for the long description so far, but I think we should update the following sentence.

```
If you have enabled the automatic backup and PITR features, the managed databases will perform back up automatically in approximately 10 seconds.
```

For example, the following.

```
If you have enabled the automatic backup and PITR features, the managed databases will perform back up automatically. And, you need to wait for approximately 10 seconds to create a sufficiently long period to avoid the clock skew issue between the client clock and the database clock.
```

What do you think?

Author: ```
**The ScalarDB or ScalarDL pod status is `Running` in the `STATUS` column.** Confirm that the pods are running and not paused.
```

Regarding the above sentence, the pause/unpause state is Scalar products original state.
So, it cannot be observed from the Kubernetes perspective.
In other words, the `STATUS` column will not be like `PAUSED` even if we pause the Scalar product pods.
It is enough to confirm the `STATUS` is `Running`.

So, I think we can remove `Confirm that the pods are running and not paused.`.

Author: In this conversation, there are two comments. Please take a look at both comments.

> 1. (DynamoDB only) You need to do additional operations to create a backup if you use the PITR feature of DynamoDB since it restores data with another name table by PITR. You should these additional steps after you complete this creating **period** steps. Please refer to the [Restore database guide](./RestoreDatabase.md#amazon-dynamodb) for more details on these steps.

Author: If we use DynamoDB, we need additional operations to create a backup since we cannot restore tables with the same name by PITR.

> 
> 1. (DynamoDB only) You need to do additional operations to create a backup if you use the PITR feature of DynamoDB since it restores data with another name table by PITR. You should these additional steps after you complete this creating **period** steps. Please refer to the [Restore database guide](./RestoreDatabase.md#amazon-dynamodb) for more details on these steps.

Reviewer 2: ```suggestion
8. **(Amazon DynamoDB only)** If you use the PITR feature in DynamoDB, you will need to perform additional steps to create a backup because the feature restores data with another name table by using PITR. For details on the additional steps after creating the exact period in which you can restore the data, please see [Restore a database in a Kubernetes environment](./RestoreDatabase.md#amazon-dynamodb).
```

> ## Multiple databases backup

Reviewer: ```suggestion
## The case for taking backups of multiple databases
```

Reviewer 2: ```suggestion
## Back up multiple databases
```

> If you use more than two databases under the Scalar products using [Multi-storage Transactions](https://github.com/scalar-labs/scalardb/blob/master/docs/multi-storage-transactions.md) or [Two-phase Commit Transactions](https://github.com/scalar-labs/scalardb/blob/master/docs/two-phase-commit-transactions.md), you must pause all Scalar products at the same time and create the same **period** for multiple databases.

Reviewer: ```suggestion
If you use more than two databases under the Scalar products using [Multi-storage Transactions](https://github.com/scalar-labs/scalardb/blob/master/docs/multi-storage-transactions.md) or [Two-phase Commit Transactions](https://github.com/scalar-labs/scalardb/blob/master/docs/two-phase-commit-transactions.md), you must pause all Scalar products together and create a period where no on-going transactions exist in the databases.
```

Reviewer 2: ```suggestion
If you are using more than two databases with ScalarDB or ScalarDL that use [Multi-storage Transactions](https://github.com/scalar-labs/scalardb/blob/master/docs/multi-storage-transactions.md) or [Two-phase Commit Transactions](https://github.com/scalar-labs/scalardb/blob/master/docs/two-phase-commit-transactions.md), you must pause all database instances on ScalarDB or ScalarDL together and create the same period where no ongoing transactions exist in the databases.
```

Author: As I mentioned in another comment, I think it would be better to describe `databases that are used by ScalarDB or ScalarDL` instead of `databases that use ScalarDB or ScalarDL`.

What do you think?

> To keep consistency between multiple databases, you must restore multiple databases to the same point-in-time data using the PITR feature.

Reviewer: ```suggestion
To keep consistency between multiple databases, you must restore multiple databases to the same point in time using a PITR feature.
```

> ## How to use scalar-admin on a Kubernetes environment

Reviewer 2: ```suggestion
## Details on using `scalar-admin`
```

### Check the Kubernetes resource name

> You must specify the SRV service URL to the `-s (--srv-service-url)` flag. In the Kubernetes environment, the format of the SRV service URL is `_my-port-name._my-port-protocol.my-svc.my-namespace.svc.cluster.local`.

Author: This `-s (--srv-service-url)` flag is an old (current) way to use `scalar-admin`. I will update this document after we will release a new tool for backup.

> You must specify the SRV service URL to the `-s (--srv-service-url)` flag. In the Kubernetes environment, the format of the SRV service URL is `_my-port-name._my-port-protocol.my-svc.my-namespace.svc.cluster.local`.

Reviewer 2: ```suggestion
You must specify the SRV service URL to the `-s (--srv-service-url)` flag. In Kubernetes environments, the format of the SRV service URL is `_my-port-name._my-port-protocol.my-svc.my-namespace.svc.cluster.local`.
```

> If you use Scalar Helm Chart to deploy Scalar products, the `my-svc` and `my-namespace` are variable depending on the environment. You must specify the Headless Service name as `my-svc` and the namespace as `my-namespace`.

Reviewer 2: ```suggestion
If you use Scalar Helm Charts to deploy ScalarDB or ScalarDL, the `my-svc` and `my-namespace` may vary depending on your environment. You must specify the headless service name as `my-svc` and the namespace as `my-namespace`.
```

    _scalardl-admin._tcp.<helm release name>-headless.<namespace>.svc.cluster.local
    ```
  * ScalarDL Auditor
    ```console
    _scalardl-auditor-admin._tcp.<helm release name>-headless.<namespace>.svc.cluster.local
    ```

The Headless Service name `<helm release name>-headless` is decided based on the helm release name. You can see the helm release name using the `helm list` command.

```console
$ helm list -n ns-scalar
NAME                    NAMESPACE       REVISION        UPDATED                                 STATUS                                                       CHART                    APP VERSION
scalardb                ns-scalar       1               2023-02-09 19:31:40.527130674 +0900 JST deployed                                                     scalardb-2.5.0           3.8.0
scalardl-auditor        ns-scalar       1               2023-02-09 19:32:03.008986045 +0900 JST deployed                                                     scalardl-audit-2.5.1     3.7.1
scalardl-ledger         ns-scalar       1               2023-02-09 19:31:53.459548418 +0900 JST deployed                                                     scalardl-4.5.1           3.7.1
```

> You can also see the Headless Service name `<helm release name>-headless` directly using the `kubectl get service` command.

Reviewer 2: ```suggestion
You can also see the headless service name `<helm release name>-headless` by running the `kubectl get service` command.
```

scalardb-headless                ClusterIP      None             <none>        60051/TCP                         2m2s
scalardb-metrics                 ClusterIP      10.111.213.194   <none>        8080/TCP                          2m2s
scalardl-auditor-envoy           LoadBalancer   10.111.141.43    <pending>     40051:31553/TCP,40052:31171/TCP   99s
scalardl-auditor-envoy-metrics   ClusterIP      10.104.245.188   <none>        9001/TCP                          99s
scalardl-auditor-headless        ClusterIP      None             <none>        40051/TCP,40053/TCP,40052/TCP     99s
scalardl-auditor-metrics         ClusterIP      10.105.119.158   <none>        8080/TCP                          99s
scalardl-ledger-envoy            LoadBalancer   10.96.239.167    <pending>     50051:32714/TCP,50052:30857/TCP   109s
scalardl-ledger-envoy-metrics    ClusterIP      10.97.204.18     <none>        9001/TCP                          109s
scalardl-ledger-headless         ClusterIP      None             <none>        50051/TCP,50053/TCP,50052/TCP     109s
scalardl-ledger-metrics          ClusterIP      10.104.216.189   <none>        8080/TCP                          109s
```

### Pause

> You can send the pause request to Scalar product pods on the Kubernetes environment.

Reviewer 2: ```suggestion
You can send a pause request to ScalarDB or ScalarDL pods in a Kubernetes environment.
```

* Example
  * ScalarDB Server
    ```console
    kubectl run scalar-admin-pause --image=ghcr.io/scalar-labs/scalar-admin:<tag> --restart=Never -it -- -c pause -s _scalardb._tcp.<helm release name>-headless.<namespace>.svc.cluster.local
    ```
  * ScalarDL Ledger
    ```console
    kubectl run scalar-admin-pause --image=ghcr.io/scalar-labs/scalar-admin:<tag> --restart=Never -it -- -c pause -s _scalardl-admin._tcp.<helm release name>-headless.<namespace>.svc.cluster.local
    ```
  * ScalarDL Auditor
    ```console
    kubectl run scalar-admin-pause --image=ghcr.io/scalar-labs/scalar-admin:<tag> --restart=Never -it -- -c pause -s _scalardl-auditor-admin._tcp.<helm release name>-headless.<namespace>.svc.cluster.local
    ```

### Unpause

> You can send the unpause request to Scalar product pods on the Kubernetes environment.

Reviewer 2: ```suggestion
You can send an unpause request to ScalarDB or ScalarDL pods in a Kubernetes environment.
```

* Example
  * ScalarDB Server
    ```console
    kubectl run scalar-admin-unpause --image=ghcr.io/scalar-labs/scalar-admin:<tag> --restart=Never -it -- -c unpause -s _scalardb._tcp.<helm release name>-headless.<namespace>.svc.cluster.local
    ```
  * ScalarDL Ledger
    ```console
    kubectl run scalar-admin-unpause --image=ghcr.io/scalar-labs/scalar-admin:<tag> --restart=Never -it -- -c unpause -s _scalardl-admin._tcp.<helm release name>-headless.<namespace>.svc.cluster.local
    ```
  * ScalarDL Auditor
    ```console
    kubectl run scalar-admin-unpause --image=ghcr.io/scalar-labs/scalar-admin:<tag> --restart=Never -it -- -c unpause -s _scalardl-auditor-admin._tcp.<helm release name>-headless.<namespace>.svc.cluster.local
    ```

> ### Check the pause completed time and unpause started time

Reviewer 2: ```suggestion
### Check the pause completed time and unpause starting time
```

Author: I used the words `pause completed` and `unpause started` based on the output of the `scalar-admin` command.

```
$ kubectl run scalar-admin-pause --image=ghcr.io/scalar-labs/scalar-admin:1.2.0 --restart=Never -it -- -c pause -s _scalardb._tcp.scalardb-headless.default.svc.cluster.local
Pause completed at 2023-03-10T05:33:55.503Z UTC (1678426435503)
```
```
$ kubectl run scalar-admin-unpause --image=ghcr.io/scalar-labs/scalar-admin:1.2.0 --restart=Never -it -- -c unpause -s _scalardb._tcp.scalardb-headless.default.svc.cluster.local
Unpause started at 2023-03-10T05:34:13.006Z UTC (1678426453006)
```

This is because users can know what `pause started time` and `unpause started time` refers to.
In other words, it is easier for users to know what values they should see, if the document's word and the command's output are the same.

However, it seems that `unpause starting` is better from the English/documentation perspective.
In this case, which word should we prioritize?
English correctness is a high priority? We should use the same word of command output?
(Or, should we update the output message of the command?)

Sorry, I don't know much about what should we do in this case from the documentation perspective.
What do you think?

> The scalar-admin pods output the **pause completed time** and **unpause started time** to the stdout. You can also see those times using the `kubectl logs` command.

Reviewer 2: ```suggestion
The `scalar-admin` pods output the pause completed time and unpause starting time to stdout. You can also see those times by running the `kubectl logs` command.
```

Author: As I mentioned in another comment, I would like to discuss whether `unpause started` or `unpause starting` is better.
